File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {
            ')': '(',
            '}': '{',
            ']': '['
        }
        for c in s:
            if c in mapping:
                if stack and stack[-1] == mapping[c]:
                    stack.pop()
                    continue
            stack.append(c)
        return not stack

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        std::stack<char> stk;
        std::unordered_map<char, char> mapping =
        {
            {')', '('},
            {'}', '{'},
            {']', '['},
        };
        for (char c : s) {
            if (mapping.contains(c)) {
                if (stk.empty() || stk.top() != mapping[c]) {
                    return false;
                }
                stk.pop();
            }
            else {
                stk.push(c);
            }
        }
        return stk.empty();
    }
};